### ISSUE TITLE
Missing event listener in Transaction

### ISSUE THREAD

**Issue by danieljwestman** [knex/knex#4085]
# Environment

Knex version: 0.21.6 - 27 September, 2020
Database + version: PG 12
OS: MacOS Catalina

The `on('start', ...)` listener is missing in the `makeTxClient` function in file `lib/transaction.js`.

See attachment.

<img width="700" alt="Skärmavbild 2020-10-27 kl  10 00 29" src="https://user-images.githubusercontent.com/20950568/97280793-0754c600-183d-11eb-9462-06ad9b0216ae.png">

Should add:

```
  trxClient.on('start', function (arg) {
    trx.emit('start', arg);
    client.emit('start', arg);
  });

```

**Comment by kibertoad** [comment]
@elhigu Should we propagate this event?

**Comment by elhigu** [comment]
Please give also an example code where `start` event is not emitting. I would also like to rename that event to something more descriptive like `before-query` or `after-query-compile`, maybe the first one is more correct describing it.

@kibertoad I'm not completely sure... I would like to see first where that gives a problem. If it prevents `start` -event from triggering for queries that are done through `trx` propagating sounds like something that should be done.

**Comment by danieljwestman** [comment]
@elhigu I need all the events to be triggered both on root and child knex instances. As it is now it is now the `start` event isn't triggered on children (transactions). the other events is working fine, as they are included in the `makeTxClient` function as above...

**Comment by danieljwestman** [comment]
If I add the snippet as above everything works like a charm (as it should in my opinion).
```

 trxClient.on('start', function (arg) {
    trx.emit('start', arg);
    client.emit('start', arg);
  });

  trxClient.on('query', function (arg) {
    trx.emit('query', arg);
    client.emit('query', arg);
  });

  trxClient.on('query-error', function (err, obj) {
    trx.emit('query-error', err, obj);
    client.emit('query-error', err, obj);
  });

  trxClient.on('query-response', function (response, obj, builder) {
    trx.emit('query-response', response, obj, builder);
    client.emit('query-response', response, obj, builder);
  });
```

**Comment by kibertoad** [comment]
@danieljwestman When is the "start" event fired?

**Comment by danieljwestman** [comment]
When I send a database query it is triggered before the query is compiled. This is important because I need to modify the knex instance. As I said, everything is working great in the root knex, and in the child knex (transaction) with the fix as above.

**Comment by elhigu** [comment]
> Please give also an example code where start event is not emitting.

With this I meant that example code, that shows exactly with what kind of initialization and knex -usecase the event is not firing.

e.g.:

```
knex.on('start', () => console.log('START FIRED'))
await knex.transaction(async trx => {
  await trx('foo').where('id', 1); // start event doesn't fire.
});
```

**Comment by danieljwestman** [comment]
It's a big Directus v9 project. Just to be sure that the problem is in the knex source I added a test.

In `knex/lib/util/make-knex.js`:

```
  // Passthrough all "start" and "query" events to the knex object.
  _addInternalListener(knex, 'start', (obj) => {

	console.log('START');

    knex.emit('start', obj);
  });
```

In the `directus/api/src/database/index.ts`:

```
database.on('start', function(builder) {

	console.log('START');

});
```

As expected the first log is triggered every time a query is made. The second one is only triggered in the root, not in transactions (because that is a new child Knex instance that doesn't inherit the start event). When I add my code snippet (to make the child instance inherit all the event types) it all works.

**Comment by elhigu** [comment]
That still really didn't answer to my request. See.  https://stackoverflow.com/help/minimal-reproducible-example https://github.com/knex/knex/blob/master/CONTRIBUTING.md#what-is-minimal-code-to-reproduce-bug-and-why-i-have-to-provide-that-when-i-can-just-tell-whats-the-problem-is

**Comment by danieljwestman** [comment]
@elhigu Okay, then I don’t know how to make it more clear... In my case it’s embedded in Directus. But, the ”problem” is pretty obvious? In the transaction you only inherit the other events, not ”start”... Right? If I add my simple lines of code, as above, everything is fine.

**Comment by elhigu** [comment]
PR with tests is welcome.

**Comment by kibertoad** [comment]
@danieljwestman I think the main part missing at this point is the actual reproduction. Not the solution, but actual runtime code that triggers the error. Do I understand correctly that executing even the simplest select statement should fire this event?

**Comment by danieljwestman** [comment]
There is no error. The event listener is just not triggered for this particular "start" event. All other event types, like "query", is triggered. This isn't strange at all since the "start" event isn't added in the make function... Why would only the other event types be added? 

Just adding the little code will fixe this issue...

**Comment by kibertoad** [comment]
@danielgindi I don't mean "error" as in an exception being thrown. Again, what I mean is I need to be able to reproduce the case of event not being propagated to transaction. Again, could you please confirm that if I do any kind of select, such as "select 1 from dual" or whatever, then "start" event will be fired for root knex instance, but not for transaction instance?
I need to know how to reproduce issue so that I could write a test for it. Thank you.

**Comment by danieljwestman** [comment]
Yes, thats correct. All queries are triggering the "start" event for "root" but none for "children" instances (transactions). The other events (like "query") are always triggered, since they are included in the transaction make funktion (as explained above).

**Comment by kibertoad** [comment]
Thank you! Will give it a shot tonight.

**Comment by danieljwestman** [comment]
Great, thanks

**Comment by danieljwestman** [comment]
@kibertoad Hi! Did it work out for you?

**Comment by kibertoad** [comment]
@danieljwestman was busy with other fixes yesterday, will focus on this one today. thank you for the ping

**Comment by kibertoad** [comment]
Released in 0.21.9
